server/path_watcher.go

```diff
@@ -0,0 +1,309 @@
+		log.Critical("%s can't recover from inconsistent cluster state on %s: notifying the current master failed: %s", watcher, event.Key, err)
+	} else if recovered {
+		watcher.updateCounter(1, "inconsistency_recovery.success")
+		log.Info("%s successfully recovered from inconsistency on %s", watcher, event.Key)
+	} else {
+		watcher.updateCounter(1, "inconsistency_recovery.not_needed")
+		log.Info("%s inconsistency on %s detected, but further events are already scheduled, the cluster will recover itself soon", watcher, event.Key)
+	}
+}
+
+func (watcher *PathWatcher) tryRecoverCompaction(ctx context.Context, event *gohan_sync.Event, compactedRevision int64) {
+	// next watch should start at lastProcessed +1, so we're setting a known-good -1
+	if err := watcher.storeRevision(ctx, compactedRevision-1); err != nil && err != errInconsistentCluster {
+		// it's not fatal: the next leader will with again fail with errCompacted and retry the recovery
+		watcher.updateCounter(1, "compaction_recovery.failed")
+		log.Warning("%s can't recover from etcd compaction on %s: %s", watcher, event.Key, err)
+	} else if err == errInconsistentCluster {
+		watcher.updateCounter(1, "compaction_recovery.not_needed")
+	} else {
+		watcher.updateCounter(1, "compaction_recovery.success")
+	}
+}
+
+func (watcher *PathWatcher) watchExtensionHandler(ctx context.Context, response *gohan_sync.Event) {
+	defer l.Panic(log)
+	for event, env := range watcher.extensions {
+		if strings.HasPrefix(response.Key, "/"+event) {
+			watcher.runExtensionOnSync(ctx, response, env.Clone())
+			return
+		}
+	}
+}
+
+// fetchStoredRevision returns the revision number stored in the sync backend for a path.
+// When it's a new in the backend, returns sync.RevisionCurrent.
+func (watcher *PathWatcher) fetchStoredRevision(ctx context.Context) int64 {
+	fromRevision := int64(gohan_sync.RevisionCurrent)
+	lastSeen, err := watcher.sync.Fetch(ctx, SyncWatchRevisionPrefix+watcher.path)
+	if err == nil {
+		inStore, err := strconv.ParseInt(lastSeen.Value, 10, 64)
+		if err == nil {
+			log.Info("%s using last seen revision `%d`", watcher, inStore)
+			fromRevision = inStore
+		} else {
+			log.Warning("%s revision `%s` is not a valid int64 number, using the current one, which is %d (%s)", watcher, lastSeen.Value, fromRevision, err)
+		}
+	} else {
+		log.Warning("%s failed to fetch last seen revision number, using the current one, which is %d: (%s)", watcher, fromRevision, err)
+	}
+	return fromRevision
+}
+
+// storeRevision puts a revision number for a path to the sync backend.
+func (watcher *PathWatcher) storeRevision(ctx context.Context, revision int64) error {
+	path := SyncWatchRevisionPrefix + watcher.path
+	value := strconv.FormatInt(revision, 10)
+	opts := make([]gohan_sync.CASCondition, 0, 1)
+	if watcher.previousProcessedRevision != int64(gohan_sync.RevisionCurrent) {
+		opts = append(opts, watcher.sync.ByValue(strconv.FormatInt(watcher.previousProcessedRevision, 10)))
```

Comment:
Shouldn't this be `ByRevision`?

Reply:
`ByRevision` isn't used anywhere outside of tests.

Reply:
so let's just remove it